### PR TITLE
Feature app data

### PR DESCRIPTION
**Pull request by davidyan74**
The reason of the large number of commits is that Tim has reapplied the old commits from DTX about PubSubWebSocketServlet to preserve the author attribution.

### CODE REVIEW

**Comment by davidyan74** [comment]
test this please

File: engine/pom.xml [hunk 1]
```diff
@@ -272,7 +272,6 @@
       <groupId>org.eclipse.jetty</groupId>
       <artifactId>jetty-websocket</artifactId>
       <version>${jetty.version}</version>
-      <scope>test</scope>
```
Comment (tweise) [inline engine/pom.xml]: Why this change?

**Comment by tweise** [comment]
Please create an APEX JIRA that summarizes the purpose of these changes.

File: engine/pom.xml (hunk 1 above)
Comment (davidyan74) [inline engine/pom.xml]: This is because previously, there was a unit test that is based on the websocket in Apex.  But now since the websocket servlet is moved to Apex, the dependency scope becomes not just test, since the websocket servlet code depends on the websocket library.
